### ISSUE TITLE
SQL DECIMAL type shouldn't be cast to float.

### ISSUE THREAD

**Issue by a-schaeffer** [bcosca/fatfree#984]
Since 3.6.0 it seems that the SQL PDO wrapper in /db/sql.php on line 372:

```php
(preg_match(
'/float|decimal|real|numeric|double/i',
$row[$val[2]])?self::PARAM_FLOAT:
\PDO::PARAM_STR)))
```

Assigns the value self::PARAM_FLOAT as the pdo_type to all columns matching the regex. Those are probably all floating point numbers indeed, however in the value() function in sql.php on line 122 the value is cast to a float! 
```php
case self::PARAM_FLOAT:
    return (float)(is_string($val)
    ? str_replace(',','.',preg_replace('/([.,])(?!\d+$)/','',$val))
    : $val);
```

F3 3.5.1 worked fine but when I upgraded to 3.6.0 my DECIMAL columns weren't returned as strings anymore and more importantly they get truncated! The DECIMAL type (and possibly others as well) should just be returned as strings, 3.6.0 is not basically undoing all the advantages that the DECIMAL type brought.

**Comment by ikkez** [comment]
at which point do they go truncated? usually a string to float cast should convert the decimal point as well, if it's properly prepared.

**Comment by a-schaeffer** [comment]
The DECIMAL type can be configured to hold values much larger than float can represent at all e.g.: DECIMAL(30,10) there is no way that can be cast to a float no matter how it is represented.
At first with smaller values everything might look just fine. Take a look at this example:

```php
$a = (float)"10000000.1";
$b = (float)"20000000.2";
$c = (float)"1234567890.1234567890";

printf("%f\r\n", $a + $b);
printf("%.10f\r\n", $a + $b);
printf("%.10f\r\n", $c);
```

And its output:

```
30000000.300000
30000000.2999999970
1234567890.1234567165
```

Also check out the big warning and example at: [http://php.net/manual/en/language.types.float.php](url) The DECIMAL type was specifically introduced to accurately represent decimal numbers. Sure, it could be that the end result of a query (while doing intermediate calculations using a query) might be fine to represent as a float eventually, but only in specific cases. Most of the time I just expect my (accurate, non-truncated) decimal string back from the database the way I stored it.

**Comment by ikkez** [comment]
Thanks for details.
>  Most of the time I just expect my (accurate, non-truncated) decimal string back from the database the way I stored it.

I'm absolutely with you. We just had some user expectations, where in case you set a float value to a mapper, that stores this value in a float-type table column, the resulting value after loading this data back again wasn't a float, but a (even broken) string. 

So, do you think this can be solved by removing `decimal` from the `preg_match` above, moving to the old behaviour and keep the current mechanics for the others?

**Comment by a-schaeffer** [comment]
I suppose if you really wanted to, you could get away with casting double/float sql types to a float in PHP. Unless you are storing large values with a lot of precision, casting a string to a float will work out. If it doesn't, the developer should've probably resorted to the DECIMAL type anyway. 

I get that perhaps some users expected an actual float to be returned if they set it as such when inserting the value. But does that really matter? After all since php is loosely typed, the following works just fine:

```php
echo "0.1" + "0.2";
```

In my opinion the SQL PDO/Mapper shouldn't do any casting at all, It should be up to the developer to decide and judge whether to cast or not. For example when encoding values to JSON, e.g.:

```php
$a = "10.125";
echo json_encode(array("this is a float" => (float)$a));
```

So bottom line is: it is probably safe-ish to keep float|double in the preg_match but all other should **definitely** be removed.

**Comment by xfra35** [comment]
Hi @a-schaeffer, thanks for reporting. We didn't think about this precision issue while fixing https://github.com/bcosca/fatfree-core/issues/106.

I agree with all what you've written, except for one thing:
> the SQL PDO/Mapper shouldn't do any casting at all

Sending a float value to database as a string *is* casting. The issue we've been trying to fix was that the float-to-string casting performed in 3.5 didn't work well with non-English locales with a comma decimal separator.

Since PHP doesn't offer a decimal type, we could add a config parameter to let the developer decide how to treat DB decimal values (either as strings or as floats).

What do you think?

**Comment by a-schaeffer** [comment]
Hi @xfra35,

> I agree with all what you've written, except for one thing:
> 
> > the SQL PDO/Mapper shouldn't do any casting at all
>
> Sending a float value to database as a string _is_ casting. 

I get what you are saying, but let my rephrase that just to be clear:
The SQL PDO/Mapper shouldn't do any casting to a type where loss of precision might occur.
Casting a float/double to a string should be fine since one can just add more bytes to the string to represent whatever float value. The opposite isn't always true because the float type only has x number of bits to work with. 

> The issue we've been trying to fix was that the float-to-string casting performed in 3.5 didn't work well with non-English locales with a comma decimal separator.

I am not sure I am fully understanding the issue with the non-English locales? Without using F3, the PHP PDO accepts floats as PDO::PARAM_STR correct? So the developer can pass an actual float type, or a string formatted according to what SQL accepts (just a decimal dot, no thousands separator or anything like that). When retrieving, a string is always returned using that exact same formatting (I think its called ROOT or US locale formatting, with just the decimal dot). Which the developer can then cast to a float (or let F3 do it, in this case using the your proposed configuration setting).
Where do non-us locales come into play? If you are talking about developers passing strings like '120.000,30' for example, then I would say the mapper shouldn't even worry about that and it is up to the developer to cast that to an actual float or properly formatted string.

> Since PHP doesn't offer a decimal type, we could add a config parameter to let the developer decide how to treat DB decimal values (either as strings or as floats).
> 
> What do you think?

Yes that would work, but please make sure to only have that option affect the SQL float and double types, never include DECIMAL/REAL/NUMERIC or similar types in that option. (i.e. remove decimal|real|numeric from the preg_match, don't even assign self::PARAM_FLOAT to those fields either, just let them be strings).

Usage of the DECIMAL/REAL/NUMERIC types is often done to work with multi precision math libraries or when when numbers with a lot of decimals points need to be stored accurately. (See the 0.1 + 0.2 example I posted earlier). No developer in their right mind would consider casting those back to a float and introducing loss of information after having just spent extra CPU cycles to achieve the precision they were looking for.

**Comment by ikkez** [comment]
hey folks, I just had to fix this now and played a while on this topic.

Just to sum this up:

Binary floating point numbers are always approximated values and have limited precision, due to the nature of the hardware they are calculated on (our x86 machines). PHP typically uses the [IEEE 754](https://en.wikipedia.org/wiki/IEEE_floating_point#Basic_formats) double precision format to calculate these numbers. The problem is that numbers represented in this format, whose denominator is not a power of two, can't be exactly represented.

I found out that the `php.ini` value `precision` is a good screw to adjust the internal float precision behaviour. It basically tells PHP how large a float can be. Yet it doesn't matter how many digits are before or after the decimal separator, it's just about the total number of digits represented:

```php
// all fine;
var_dump( 12345678912.3 );	// float 12345678912.3
var_dump( 1234567891.23 );	// float 1234567891.23
var_dump( 123456.789123 );	// float 123456.789123
var_dump( 1.23456789123 );	// float 1.23456789123

// truncated:
var_dump( 1234567890.123456 );	// float 1234567890.12
var_dump( 12.34567890123456 );	// float 12.3456789012

// adjusted precision and all fine again:
ini_set('precision',16);
var_dump( 1234567890.123456 );	// float 1234567890.123456
var_dump( 12.34567890123456 );	// float 12.34567890123456
```

This is basically the reason why @a-schaeffer got truncated values from the database, because the column data types for decimal fields can represent a much higher precision than PHP float is able to create by default settings.

Nevertheless, since the final precision is limited, there are some thinks to consider when the `precision` is increased:

* float values get extrapolated at a certain precision, mostly between 15-18 and greater.  in the float value

	```php
	ini_set('precision',14);
	var_dump(1234.56 - 1233.03); // float 1.53
	ini_set('precision',15);
	var_dump(1234.56 - 1233.03); // float 1.52999999999997
	```

	results depend on the mathematical operations being made, because they may resolve in different rounding

	```php
	$a = 10000000.1;
	$b = 20000000.2;
	ini_set('precision',16);
	var_dump($a + $b); // float 30000000.3
	ini_set('precision',17);
	var_dump($a + $b); // float 30000000.299999997
	```

	independent of precision

	```php
	var_dump(number_format(0.1 + 0.2,16,'.','')); // string '0.3000000000000000' (length=18)
	var_dump(number_format(0.1 + 0.2,17,'.','')); // string '0.30000000000000004' (length=19)
	```

* values lose precision due to rounding in the order of *1.11e-16*

	```php
	ini_set('precision',12);
	var_dump( 1234567890.1234567890 ); // float 1234567890.12
	ini_set('precision',16);
	var_dump( 1234567890.1234567890 ); // float 1234567890.123457
	ini_set('precision',18);
	var_dump( 1234567890.1234567890 ); // float 1234567890.12345672 <-- ?
	```


**Alright, and now?**

Well the after investigating on how the float conversion might take apart your whole application, I thought it could be an option to take that
precision value into account. I am totally with @a-schaeffer that our recently added float conversion should not truncate the data.

To draw some conclusions:

1. We should not type-convert numeric values to float, if the value is not representable by float (current system `precision`)
2. Higher `precision` values might introduce rounding errors, so keep it `<=15`
3. The rounding errors we're talking about are just caused by the calculated representation of the float, which finally means that the way how to
display a float is essential:

	```php
	ini_set('precision', 20); // set to 20
	var_dump( $a - $b );
	var_dump( floatval(round($a - $b, 2)) );
	var_dump( number_format($a - $b, 2) );
	var_dump( bcsub($a, $b, 2) );
	// results:
	// float 584.15999999997438863
	// float 584.15999999999996817
	// string '584.16' (length=6)
	// string '584.15' (length=6)
	```

	So best option is to use `number_format` wherever you want to put down a big float as intentionally formatted string. The f3->format method suits as well.


I've added a fix now, that skips the float conversion when the precision is lower than the decimal string length, which should be a good compromise.
Additionally you can always set a custom data type for each field to skip the whole part intentionally like this:

```php
class Product extends \DB\SQL\Mapper {
	function __construct($db,$table,$fields=NULL,$ttl=60) {
		parent::__construct($db,$table,$fields,$ttl);
		$this->fields['price']['pdo_type']=\PDO::PARAM_STR;
	}
}
```

---

The good news about this is that it's more or less fully automated now. The values on the mapper are float if possible and (non-truncated) string otherwise.

So a snippet like this
```php
echo json_encode(array("this is a float" => (float)$a));
```
which could truncate your data on the last road to the final response is not needed anymore.

thoughts?

some references: [[1]](http://stackoverflow.com/questions/14587290/can-i-rely-on-php-php-ini-precision-workaround-for-floating-point-issue) [[2]](http://stackoverflow.com/questions/35111302/round-breaks-when-php-precision-is-set-to-18?rq=1) [[3]](http://stackoverflow.com/questions/588004/is-floating-point-math-broken)

**Comment by xfra35** [comment]
Great work @ikkez 👍

Thanks for digging into this "precision hell" ^^

If I understand well the changes introduced in https://github.com/bcosca/fatfree-core/commit/e647bbd76b09580ef21db571c81034926901c950, the returned type depends on the number of digits. Doesn't this make it difficult to predict the output? One decimal less and you get a `float` where you expected a decimal:

* `value('1234567890.123')` => string
* `value('234567890.123')` => float

**Comment by ikkez** [comment]
Well maybe we should write a little guide about float expectation and behaviour. Initially you came up with the "issue" that db-float columns are not representatad as float in php. Now it does. Sure it depends somehow on the environment, but I think with the most default php settings you probably have no problems with geopositional data (`53.1234567`) or webshop-like financial data at a reasonable value. For any applications that have to deal with bigger numbers or precision, you cannot rely on float or string and the default math operations at all and should probably use the [BC Math lib](http://php.net/manual/en/ref.bc.php) and custom output formatting (`number_format`). In that case it doesn't matter if the float becomes a string, as this is no problem for bc math or number_format, and is just made to keep the precision while storing into the database. We cannot use float for representation nor calculation for high precision apps and the developer should care about this. This type conversion doesn't matter at this point, since the application should be prepared for this precision calculation and representation anyhow:

```php
var_dump(100200300.123456);
var_dump(number_format(100200300.123456,6,'.',''));
// float 100200300.123
// string '100200300.123456' (length=16)

var_dump(100200300.123456+100200300.123456);
var_dump('100200300.123456'+'100200300.123456');
// float 200400600.247
// float 200400600.247

var_dump(number_format(100200300.123456+100200300.123456,6,'.',''));
var_dump(number_format('100200300.123456'+'100200300.123456',6,'.',''));
var_dump(bcadd('100200300.123456','100200300.123456',6));
// string '200400600.246912' (length=16)
// string '200400600.246912' (length=16)
// string '200400600.246912' (length=16)
```
if you think that people might feel uncomfortable with this, maybe we make the "float" detection preg_match configurable, but I think the most confusion part is more about how to echo a high presicion float value that matches the expectations, which ultimately answer is to use`number_format`.

**Comment by xfra35** [comment]
My concern is (again) about locales that have comma decimal separator.

With the change introduced in https://github.com/bcosca/fatfree-core/commit/e647bbd76b09580ef21db571c81034926901c950, we take the risk to lose the decimal part (because of the comma) or even the whole number:

```php
echo $mapper->num1; // 1234567890.123 (string)
echo $mapper->num2; // 234567890.123 (float)
setlocale(LC_ALL,'en_US.UTF-8');
echo bcadd($mapper->num1,(string)$mapper->num2);// 1469135780.246 (OK)
setlocale(LC_ALL,'fr_FR.UTF-8');
echo bcadd($mapper->num1,(string)$mapper->num2);// 1234567890.123 (!)
// (num2 is ignored because of the comma)
```

I think it's safer to let the developer decide how to handle the `PARAM_FLOAT`: to cast everything to float, or to cast nothing at all.

**Comment by a-schaeffer** [comment]
Hey all, It's probably best to not over complicate this issue.
If I am correct the current solution as committed by @ikkez still includes types other than the SQL FLOAT and DOUBLE types. e.g.: a DECIMAL type can get possibly cast to a float if its length is <= the current precision setting, even though no precision is lost in this case it shouldn't be happening.

The F3 framework should assume that anybody using a type other than FLOAT or DOUBLE  did so with the knowledge that for their specific application the limited range and accuracy of the FLOAT and DOUBLE types weren't sufficient. I don't expect any framework or library to mess around with my numbers when I implicitly indicated (by not using FLOAT/DOUBLE) that a certain precision is important to me.

I'd suggest to only assign PARAM_FLOAT to FLOAT and DOUBLE types. i.e.:

```php
(preg_match(
'/float|double/i',
$row[$val[2]])?self::PARAM_FLOAT:
\PDO::PARAM_STR)))
```

This will keep developers that expect _float in -> float out_ behavior happy, and developers that have expressed their desire to work with higher precision values will get what they want as well.

**Comment by xfra35** [comment]
Great 👍 

Can we revert https://github.com/bcosca/fatfree-core/commit/e647bbd76b09580ef21db571c81034926901c950, as it opens the door to very tricky bugs (cf. comment above)?

**Comment by ikkez** [comment]
🤔  When we revert that, you could loose data in CRUD scenarios, when you aren't aware of the data size and the chosen db data type. When using float, we should not cast it to string ourself, but work with the float values or use number_format to get a string representation.
Well, if that is the consequence, I wonder myself what issue we have solved by adding the whole float-support to the mapper. Just for the sake of completness? We're not really able to work with float in a natural way because of so many pitfalls:
*  we may not cast it to string in multi-lang apps, because the comma-separator can lead to unpredicatable issues 
*  precision could get lost if the app is not designed from the ground up to compensate this
*  to display floats in a predicatable manner, we need to format those back to string anyways (with `number_format`)

So what did we gained by casting to float? I start thinking that it maybe was a bad idea to add this and we should just remove that and keep the fix for bcosca/fatfree-core#106

**Comment by xfra35** [comment]
Well here's the rationale:
* for geopositional data (53.1234567) or webshop-like financial data, store data in db as `FLOAT` and manipulate this data as a PHP `float` inside your mappers, no matter the selected locale
* if you're using English locale only, you can as well store the data as `DECIMAL`, you won't notice the difference (float<->string cast is transparent)
* if you're storing high precision numbers, make sure the php.ini `precision` parameter can handle that. Otherwise, store data as `DECIMAL`, in which case you should preferably stick to English locale.

**Comment by a-schaeffer** [comment]
I think the whole locale issue should be left out of the mapper.
The developer should always present FLOAT/DOUBLEs to the mapper as a PHP `float` or a string formatted according to the "root" locale (i.e. "en-US" using the decimal dot and nothing else).
The mapper should then always output a PHP `float` for SQL FLOAT/DOUBLE columns.
Any other SQL values like NUMERIC/DECIMAL/REAL should be treated as discussed before, they should be just opaque strings to the mapper.

Formatting is then up to the developer, which can be done using PHP's built in functions or some locale dependent helper function.

**Comment by xfra35** [comment]
> I think the whole locale issue should be left out of the mapper

That would be nice, but unfortunately you can't ignore this issue. Because when you assign `12.34` to a mapper, this `12.34` has to be converted to a string (`PDO::PARAM_STR`), and that conversion can introduce a comma, depending on the locale:

```php
echo (string)12.34;// 12,34
```

**Comment by ikkez** [comment]
Actually, there doesn't seem to be any issue with binding a float with `PARAM_STR`:

```php
setlocale(LC_ALL,'fr_FR.UTF-8');
$mapper->load();
$mapper->price_float = 27521.89;
$mapper->save();
var_dump($mapper->price_float); // float 27521.89
$mapper->reset();
$mapper->load();
var_dump($mapper->price_float); // string '27521.89'
```

**Comment by a-schaeffer** [comment]
> > I think the whole locale issue should be left out of the mapper
> 
> That would be nice, but unfortunately you can't ignore this issue. Because when you assign 12.34 to a mapper, this 12.34 has to be converted to a string (PDO::PARAM_STR), and that conversion can introduce a comma, depending on the locale:
> 
> ```php
> echo (string)12.34;// 12,34
>```

What about using sprintf with the %F non-locale aware format specifier.
```php
var_dump(rtrim(sprintf('%F', 12.34), '0')); // string(5) "12.34"
```

This would ensure that all strings are now only using the decimal dot, regardless of locale. A call to floatval() can then be used by the mapper to respect the _float in -> float out_ behavior desired by some users.

**Comment by ikkez** [comment]
printf causes data loss, you've shown us in https://github.com/bcosca/fatfree/issues/984#issuecomment-269010059
On the current commit, there's no problem with storing string or float values to a float/double field regardless of the locale

**Comment by a-schaeffer** [comment]
> printf causes data loss, you've shown us in #984 (comment)
> On the current commit, there's no problem with storing string or float values to a float/double field regardless of the locale.

It's the other way around, I wrote that comment to make it clear to other developers that floating point variables can't be trusted to accurately represent every decimal number. To demonstrate that casting DECIMAL values to floats was a bad idea.
Casting using `(string)` isn't any less (in)accurate than using `sprintf/printf`.

**Comment by xfra35** [comment]
@a-schaeffer I didn't know about that non-locale aware `%F`. Interesting!

@ikkez, you've written:
> I start thinking that it maybe was a bad idea to add this

I start thinking that I'd rather store prices in cents and forget about floats ^^

So let's sum it up. We have two issues: a locale issue and a precision issue. The locale issue (raised in bcosca/fatfree-core#106) was initially due to [that line](https://github.com/bcosca/fatfree-core/blob/e33c7a65d9006851cd80d693a07d9d2b5f94e53a/db/sql.php#L138) introducing a comma. The precision issue is due to the float casting.

So let's fix that comma issue without casting to float and we'll be set. I've created a PR for that.
